### PR TITLE
Fix "illegal operation on a directory, realpath" on RAM Disk

### PR DESCRIPTION
**Pull request by duykhang53**
The native method of `fs.realpath()` cause error on RAM Disk created by imDisk.

Angular developers might get this error when running `ng s` on RAM Disk when they using scss in their project

From tracing in **postcss-import**, I got *EISDIR: illegal operation on a directory, realpath 'R:\angular-project'*.
image

For the fix, ` fs.realpath.native()` will fallback to `fs.realpath()` if *errno* is illegal operation(-4068).

### CODE REVIEW

File: lib/async.js [hunk 1]
```diff
@@ -5,7 +5,19 @@ var nodeModulesPaths = require('./node-modules-paths.js');
...
-var realpathFS = fs.realpath && typeof fs.realpath.native === 'function' ? fs.realpath.native : fs.realpath;
+function realpathFS(path, callback) {
+    if (typeof fs.realpath.native === 'function') {
+        fs.realpath.native(path, function (realpathErr, realPath) {
+            if (realpathErr && realpathErr.errno === -4068) {
+                fs.realpath(path, callback);
+            } else {
+                return callback(realpathErr, realPath);
+            }
+        });
```
Comment (ljharb) [inline lib/async.js:16]: this means we pay a performance cost every time. is this necessary, or on a given system where it fails, will it always fail?

**Review by ljharb — CHANGES_REQUESTED** [review]
If we need this complexity, let's please add a regression test.

File: lib/async.js (hunk 1 above)
Comment (duykhang53) [reply lib/async.js:16]: This code is designated for the `postcss-import` on RAM Disk. This means the I/O speed is significantly improved. The production build of my Angular project takes 60-120 secs with `fs.realpath.native()` on my 2TB Seagate FireCuda SSHD but it only takes ~60 secs on RAM Disk mounted by imDisk with `fs.realpath()` (*where `fs.realpath.native()` is illegal operation*). So in general, no performance costs are paid.

Comment (ljharb) [reply lib/async.js:16]: I'm not sure how that's possible - this is always calling realpath.native, and when it errors, is *adding* the cost of calling realpath.

(to be clear, I have no idea what imDisk is, and the last time I heard anything about RAM Disks was in the late 1990s)

Comment (duykhang53) [reply lib/async.js:16]: imDisk is a software to create a virtual disk using RAM on Windows: https://imdisk.en.lo4d.com/windows 

The cost of calling is the trade-off for I/O speed and it only adding the call on illegal operation error.

You can clone this repository using Windows on both RAM Disk created by imDisk and your HDD/SSHD/SSD https://github.com/scttcper/ngx-toastr. After the clone, run ```npm i && npm start```. On RAM Disk, the server will fail and you will get 
image

The cause is from `postcss-import` using this repository to resolve the imported files from *node_modules* in *src/style.scss*. So if you alter *./node_modules/resolve/lib/async.js*, the error is gone.

Comment (duykhang53) [reply lib/async.js:16]: > imDisk is a software to create a virtual disk using RAM on Windows: https://imdisk.en.lo4d.com/windows
> 
> The cost of calling is the trade-off for I/O speed and it only adding the call on illegal operation error.
> 
> You can clone this repository using Windows on both RAM Disk created by imDisk and your HDD/SSHD/SSD https://github.com/scttcper/ngx-toastr. After the clone, run `npm i && npm start`. On RAM Disk, the server will fail and you will get
> image
> 
> The cause is from `postcss-import` using this repository to resolve the imported files from _node_modules_ in _src/style.scss_. So if you alter _./node_modules/resolve/lib/async.js_, the error is gone.

It's ```npm i ; npm start``` on PowerShell.

Comment (ljharb) [reply lib/async.js:16]: I'm concerned about:
1. adding a fix without a regression test, because there's no protection from breaking it again
1. adding a complicated fix that might slow down the majority case to service a *very* niche edge case

**Comment by SimenB** [comment]
I wonder if node can do something to improve the native realpath. The promise api uses the native implementation under the hood (https://nodejs.org/api/fs.html#fs_fspromises_realpath_path_options), so it'll probably become more prevalent

**Comment by duykhang53** [comment]
@SimenB promise APIs don't do the trick, EISDIR still occurred

image

**Comment by SimenB** [comment]
@duykhang53 that makes sense - it's the same API. My point was that it shouldn't be up to consumers to deal with these issues, it should be fixed upstream in Node itself.

**Comment by duykhang53** [comment]
This issue occurred since node v6 and now it's v14.
https://github.com/nodejs/node/issues/6861

I think it's better to apply a fallback rather than wait for nodejs to fix it.

Cost of calling just slightly affected the performance but that would be better to make the code run instead of throwing errors.

**Comment by ljharb** [comment]
@duykhang53 the last comment in that issue, https://github.com/nodejs/node/issues/6861#issuecomment-233126946, suggests that the proper solution is to abandom imdisk in favor of `aim_ll`.

**Comment by last-Programmer** [comment]
you can use osfmount and mount as physical drive and it works out of the box.
